File: FunctionalityTest/FunctionalityTests.R
# Function tests for urban ABM

library(tidyverse)
library(ggnewscale)
library(patchwork)
# library(gifski)
# library(gganimate)
library(sf)

setwd("./FunctionalityTest")

# Movement/distance decay test -------------
raccoons <- read.csv("mvt_test.csv") %>%
  filter(rep == sample(1:max(rep), 1))

hr_props <- raccoons %>%
  filter(week >= 27 & week <= 40) %>%
  group_by(id, x, y) %>%
  summarise(count = n()) %>%
  ungroup() %>%
  group_by(id) %>%
  mutate(prop = count/sum(count)) %>%
  filter(prop > 0.1) %>%
  ungroup()
  
hr_sizes <-hr_props %>%
  dplyr::select(id, x, y) %>%
  group_by(id, x, y) %>%
  distinct() %>%
  st_as_sf(coords = c("y", "x")) %>%
  ungroup() %>%
  group_by(id) %>%
  summarize(geometry = st_union(geometry)) %>%
  st_buffer(dist = 0.5) %>%
  # st_concave_hull(ratio = 0.5, allow_holes = F)
  st_convex_hull()
  
summary(st_area(hr_sizes)/4)

# Look at land cover-------------
# Load in data
landscape <- read.csv(file='ExampleLand.csv', header = F)

# Convert to long format
df <- landscape %>%
  as.data.frame(.) %>%
  mutate(y = 1:nrow(landscape)) %>%
  pivot_longer(cols = -y, names_to = 'x', values_to = 'hab') %>%
  mutate(x = rep(1:ncol(landscape), nrow(landscape))) %>%
  mutate(hab = as.character(hab)) %>%
  mutate(hab = factor(hab, levels = c(1:7, 10)))

levels(df$hab) = c("Deciduous", "DevLo", "Pasture",
                                 "DevHi", "Wetlands", "Conifers",
                                 "Crops", "Buffer")

ggplot(data = df, aes(x=x, y=y, fill=hab))+
  geom_tile()+
  scale_fill_viridis_d(name = "Habitat", begin = 1, end = 0)+
  theme(axis.title = element_blank(), axis.text = element_blank(),
        legend.position = "None")
        #legend.key.size = unit(0.2, 'in'))

# ggsave(filename = "sample_landscape.jpeg", height = 2, width = 4,
#        units = "in")

# Length of latent period ------------------
dis <- read.csv("disease.csv")

ggplot(data = dis, aes(x = time_since_inf))+
  geom_bar(fill = 'lightgray', color = 'black')+
  labs(x = "Length of latent period (Weeks)", y = "Count")+
  theme_bw(base_size = 12)+
  theme(panel.grid = element_blank())

ggsave(filename = "latent_period.jpeg", height = 3, width = 4,
       units = "in")

# Recovery probability -----------
rec <- read.csv("disease.csv")

rec2 <- rec %>%
  group_by(week) %>%
  summarise(n_transition = n(), n_recover = min(recover)) %>%
  ungroup() %>%
  summarise(perc = sum(n_recover)/sum(n_transition))

rec2

# Mortality tests --------------------
mort_counts <- read.csv(file = "mvt_mortality_test.csv")

# Mortality as % of population
perc_df <- mort_counts %>%
  mutate(year = c(rep(1, 52), rep(2,52))) %>%
  rowwise(step) %>%
  mutate(perc_deaths = sum(n_random_mort, n_dis_mort, orphan_mort,
                           juvie_cc_mort, adult_cc_mort)/
           pop_size)

summary(perc_df$perc_deaths)
summary(perc_df$perc_deaths[perc_df$year==1])
summary(perc_df$perc_deaths[perc_df$year==2])

ggplot(perc_df, aes(x = step, y = perc_deaths, 
                    color = factor(year)))+
  geom_point(size = 2)+
  scale_color_manual(values = c("black", "limegreen"),
                     name = "Year")+
  labs(x = "Week", y = "Mortality Rate")+
  theme_bw(base_size = 13)+
  theme(panel.grid = element_blank())

ggsave(filename="mortality_Test.jpeg", height = 3, width = 6,
       units = "in")

# Causes of mortality
more_percs <- mort_counts %>%
  mutate(year = c(rep(1, 52), rep(2,52))) %>%
  rowwise(step) %>%
  mutate(total_deaths = sum(n_random_mort, n_dis_mort, orphan_mort,
                            juvie_cc_mort, adult_cc_mort)) %>%
  mutate(across(n_random_mort:adult_cc_mort, 
                .fns = ~ .x/total_deaths,
                .names = "{.col}_perc"))

no_dis <- more_percs %>%
  filter(year == 1) %>%
  select(n_random_mort_perc:adult_cc_mort_perc) #%>%
  # filter(step > 20 & step < 40)

summary(no_dis$juvie_cc_mort_perc)
colMeans(no_dis, na.rm = T)

dis <- more_percs %>%
  filter(year == 2) %>%
  select(n_random_mort_perc:adult_cc_mort_perc)

colMeans(dis, na.rm = T)

# Dispersal tests --------------------------------------
disp <- read.csv('disp_test.csv')

# Filter to get start and end points of each raccoon
disp.points <- disp %>%
  group_by(id) %>%
  filter(week == min(week) | week == max(week)) %>%
  arrange(id) %>%
  distinct() %>%
  mutate(count = n())

# Calculate total displacement in km
mvts <- disp.points %>%
  filter(count > 1) %>%
  mutate(step = case_when(week == min(week) ~ 1, 
                          TRUE ~ 2)) %>%
  select(-c(week, count)) %>%
  pivot_wider(names_from = step, values_from = c(x,y)) %>%
  mutate(dist = sqrt((x_2-x_1)^2 + (y_2-y_1)^2)*0.5)

# Adults
a.mvt <- mvts %>%
  filter(age == "A") %>%
  pull(dist)
# some didn't move
a.0 <- length(which(disp.points$age == "A" & disp.points$count == 1))
a.mvt <- c(a.mvt, rep(0, a.0))
hist(a.mvt)
summary(a.mvt)

a.plt <- ggplot(mapping=aes(a.mvt))+
  geom_histogram(binwidth = 0.5, color = 'black', fill = 'lightgray')+
  labs(x = "Annual Displacement (km)", y = "Frequency")+
  scale_x_continuous(breaks = seq(0,5,1))+
  theme_bw(base_size = 14)+
  theme(panel.grid = element_blank())

# all juveniles moved in this test
j.mvt <- mvts %>%
  filter(age == "J") %>%
  pull(dist)
hist(j.mvt)
summary(j.mvt)

j.plt <- ggplot(mapping=aes(j.mvt))+
  geom_histogram(binwidth = 0.5, color = 'black', fill = 'lightgray')+
  labs(x = "Annual Displacement (km)", y = "Frequency")+
  scale_x_continuous(breaks = seq(0,7,1))+
  theme_bw(base_size = 14)+
  theme(panel.grid = element_blank())

(a.plt | j.plt) +
  plot_annotation(tag_levels = "a")

File: FunctionalityTest/FunctionalityTests.jl
using DataFrames
using StatsBase
using Distributions
using Random
using NeutralLandscapes 
using CSV

# Land proportions calculated from Burlington raster data
land_proportions =  [0.2585, 0.2337, 0.1915, 0.1266, 0.0899, 0.0619, 0.0267, 0.0079, 0, 0]

# Data frame of habitat types & movement coefficients (see McClure et al. 2022)
hab_names = ["Deciduous", "DevLo", "Pasture", "DevHi", "Wetlands", "Conifers", "Crops", "Shrub",
                "barrier", "buffer"]
hab_coefs = [0.124, 0, -0.044, -0.496, 0.56, -0.143, -0.556, 0.441, 0, -0.5]
hab_frame = DataFrame(type = hab_names, prop = land_proportions, coef = hab_coefs)

# Load in functions
include("Functions_smol.jl")

# Define parameters
Params = DataFrame(seros = 0.0, imm_rate = 0, imm_dis = 0, imm_type = "wave")

# Simulation function
function the_mega_loop(;years, time_steps, seros, rep, immigration_type, immigration_disease, immigration_rate, 
                        outputs, movement_test=false, mortality_test=false)

    # create landscape
    if movement_test==true
        land_size=20
    else
        land_size=60
    end
    
    landscape = initialize_land(land_size=land_size, habitats = hab_frame, movement_test=movement_test)

    # Populate landscape
    lil_guys = populate_landscape(seros=0, guy_density=0.5, movement_test=movement_test)

    # define home coordinates for distance-decay function
    home_coords = deepcopy(lil_guys[:,[1,2,3]])

    for year in 1:years
        for step in 1:time_steps
            # Initialize disease at year 2, when population stabilizes
            if movement_test==false
                if year == 2 && step == 1
                    initialize_disease(lil_guys)
                end
            end

            # Lots of death
            dont_fear_the_reaper(dat=lil_guys, home=home_coords, step=step, mortality_test=mortality_test)

            # Move around
            moves = look_around.(lil_guys.x, lil_guys.y, land_size)
            move(moves, lil_guys, home_coords, landscape, 500, -0.001)

            if movement_test==false
                # Spread disease
                spread_disease(dat=lil_guys, home=home_coords)
            end

            # Reproduction occurs at specific time steps
            if step == 18
                reproduce(lil_guys, home_coords)
            end
        
            # Dispersal
            if step == 43
                # all juveniles go through the dispersal function, but a dispersal distance of 0 is possible
                juvies_leave(lil_guys, home_coords, land_size)
                        
                # Not all adults affected by this function, and some have a dispersal distance of 0
                adults_move(lil_guys, home_coords, land_size, year)
            end

            # Function where some infected guys become symptomatic or recover
            change_state(lil_guys)

            # all guys age 1 week
            lil_guys.age = lil_guys.age .+ 1

            # Update time since infection & disease
            lil_guys.time_since_inf[lil_guys.incubation .== 1] = lil_guys.time_since_inf[lil_guys.incubation.==1] .+ 1
            lil_guys.time_since_disease[lil_guys.infectious .== 1] = lil_guys.time_since_disease[lil_guys.infectious.==1] .+ 1

            if movement_test==false
                elimination = ifelse(sum(lil_guys.incubation) .== 0 .&& sum(lil_guys.infectious) .== 0, "True", "False")
            end

            # Code for testing movement:
            if movement_test==true
                if year > 1 && step < 43
                    df_step = DataFrame(rep=rep, year=year, week=step, id=lil_guys.id, x=lil_guys.x, y=lil_guys.y)
                    append!(outputs, df_step, promote = true)
                end
            end
            
            # Code for testing disease transmission:
            if movement_test==false
                if year >= 2
                    # get locations of symptomatic guys
                    infec = filter(:incubation => x -> x .== 1, lil_guys)

                    # put it in a data frame
                    frame = DataFrame(year = year, week = step, id = infec.id, x = infec.x, y = infec.y,
                                            inc = infec.incubation, inf = infec.infectious)

                    # Calculate summary statistics and append to data frame
                    append!(outputs, frame, promote = true)
                end
            end

            # Code for testing mortality
            if mortality_test==true
                # Code for adding rows to data frame
            end
        end
    end
    # Include this line to save the landscape:
    #CSV.write("ExampleLand.csv",  Tables.table(landscape), writeheader=false)
end

# Run it!
#=
# Data frame for movement tests
outputs = DataFrame([[],[],[],[],[],[]],
                    ["rep", "year", "week", "id", "x", "y"])

# Run it
reps = 5
for rep in 1:reps
    the_mega_loop(years=2, time_steps = 52, seros=Params[!,1][1], rep=rep, immigration_disease = Params[!,3][1], 
                    immigration_type=Params[!,4][1], immigration_rate = Params[!,2][1], outputs = outputs,
                    movement_test=true)
end

# Create filename
filename = "mvt_test.csv"
=#

# Data frame for disease transmission tests
outputs = DataFrame([[], [], [], [], [], [], []], 
                    ["year", "week", "id", "x", "y", "inc", "inf"])

                    

# Data frame for mortality tests
#=
outputs = DataFrame([[], [], [], [], [], [],[]], 
            ["step", "n_random_mort", "n_dis_mort", "orphan_mort", "juvie_cc_mort", "adult_cc_mort", "pop_size"])

reps = 1

for rep in 1:reps
    the_mega_loop(years=2, time_steps = 52, seros=Params[!,1][1], rep=rep, immigration_disease = Params[!,3][1], 
                        immigration_type=Params[!,4][1], immigration_rate = Params[!,2][1], outputs = outputs)
end

# Create filename
filename = "mvt_mortality_test.csv"
=#

# Save results
CSV.write(filename, outputs)
